proposals/add-deque-struct.md

```diff
@@ -0,0 +1,153 @@
+# Proposal: Add a `Deque` struct to the stdlib
+
+## Goals
+
+The goals of this document are
+
+1) Explain what problems we want to solve with `Deque`, and what problems we
+   don't want to solve.
+2) Explain how the internal storage of this double-ended
+   queue is managed. Give explored alternatives.
+3) Propose an API which can be used as a "todo list"
+   for contributors to implement.
+
+## 1 - Why we want to add a `Deque` struct to the stdlib
+
+All major languages have a double-ended queue in their stdlib.
+And they are useful for a number of reasons.
+The major one is that it allows users to work with
+FIFO (first-in first-out) and LIFO (last-in first-out) queues,
+which are very common in programming.
+
+For context, we are still at the start of Mojo's stdlib.
+Given the number of maintainers,
+we want to create a new data structure which
+doesn't add too much code and complexity to the codebase.
+
+With those constraints we give the following goals:
+
+1) Provide a data structure which allows users to
+   implement a fast FIFO and LIFO
+2) Has a familiar api with already existing Mojo
+   collections (`List`, `InlineList`, `String`, `VariadicList`, ...)
+3) Has a familiar api to Python users
+4) Can be efficiently converted from and into a `List`.
+5) Has fast access to its elements (cache-friendly).
+
+The non-goals are:
+
+1) Fast insert in the middle of
+   the `Deque` (motivation for this use case is not obvious)
+2) Thread-safety (require a certain amount of complexity)
+
+## 2 - Internal storage
+
+`Deque` will allocate a buffer of capacity `capacity` and
+use a similar algorithm as `List` to make it grow when necessary.
+
+To avoid reallocation when doing frequent append
+and pop on different sides of the
+`Deque`, which is common for FIFO queues,
+we'll use this buffer as a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer).
+
+This is how Rust's
+[`VecDeque`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html)
+is implemented. Note that it differs from
+[Python's internal implementation of `deque`](https://www.laurentluce.com/posts/python-deque-implementation/),
+as Python's implementation uses multiple blocks of memory
+with pointers between each other.
+
+Note that Python's implementation of `Deque` could be considered in the future but
+is prohibilively complexe right now given
+the number of maintainers and dynamism of the codebase.
+
+### Tradeoffs between the Rust and Python implementation
+
+**Rust**:
+
+- Efficient indexing
+- Cache friendly
+- Not thread safe
+- Needs to move all existing elements when growing the buffer, O(N) worst case.
```

Comment:
This would mean that the contents need to be `Movable`. This could be a reasonable restriction in the short term, but eventually queues of non-movable types should be supported as well (when we enable in-place construction).